### PR TITLE
SLCORE-429 Manage the embedded server from the new backend

### PR DESCRIPTION
**Pull request by damien-urruty-sonarsource**

### CODE REVIEW

**Review by henryju — CHANGES_REQUESTED** [review]
Some comments, but most of it LGTM. I'm eager to see it used in all IDEs :)

File: client-api/src/main/java/org/sonarsource/sonarlint/core/clientapi/SonarLintClient.java [hunk 1]
```diff
@@ -42,11 +43,12 @@ public interface SonarLintClient {
 
   /**
    * Temporary workaround until we decide what to do regarding HTTP requests
+   * @param connectionId nullable when we need an unauthenticated http client
    * @deprecated will be removed
    */
   @Deprecated(forRemoval = true)
   @CheckForNull
-  HttpClient getHttpClient(String connectionId);
+  HttpClient getHttpClient(@Nullable String connectionId);
```
[the remark this answers is not in the document]
Comment (damien-urruty-sonarsource) [reply client-api/src/main/java/org/sonarsource/sonarlint/core/clientapi/SonarLintClient.java:55]: Yes good idea

File: core/src/main/java/org/sonarsource/sonarlint/core/AuthenticationHelperServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+  public void initialize(String ideName) {
+    this.ideName = ideName;
+  }
+
+  @Override
+  public CompletableFuture<HelpGenerateUserTokenResponse> helpGenerateUserToken(HelpGenerateUserTokenParams params) {
+    var futureTokenResponse = new CompletableFuture<HelpGenerateUserTokenResponse>();
+
+    // go async to release the current thread
+    CompletableFuture.runAsync(() -> {
```
[the remark this answers is not in the document]
Comment (damien-urruty-sonarsource) [reply core/src/main/java/org/sonarsource/sonarlint/core/AuthenticationHelperServiceImpl.java:60]: I don't think using `supplyAsync` would help here, this method takes a Supplier, and we wouldn't not be able to supply the token immediately. That's why I create a separate future that will be resolved later when we receive the token

File: core/src/main/java/org/sonarsource/sonarlint/core/embedded/server/AwaitingUserTokenFutureRepository.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import org.sonarsource.sonarlint.core.clientapi.authentication.HelpGenerateUserTokenResponse;
+
+public class AwaitingUserTokenFutureRepository {
+  private Optional<CompletableFuture<HelpGenerateUserTokenResponse>> futureResponse = Optional.empty();
+
+  public void setFutureResponse(CompletableFuture<HelpGenerateUserTokenResponse> futureResponse) {
+    this.futureResponse = Optional.of(futureResponse);
```
[the remark this answers is not in the document]
Comment (damien-urruty-sonarsource) [reply core/src/main/java/org/sonarsource/sonarlint/core/embedded/server/AwaitingUserTokenFutureRepository.java:30]: Probably setting a new future should cancel a previous one. And yes for the reference

**Review by henryju — APPROVED** [review]
Well done! LGTM!

**Review by henryju — APPROVED** [review]

**Pull request merged**